ddspp.h

```diff
@@ -1005,32 +1005,38 @@ namespace ddspp
 
 		if (desc.type == Texture3D)
 		{
-			for (unsigned int m = 0; m < mip; ++m)
+			for (unsigned int m = 0; m <= mip; ++m)
 			{
-				unsigned long long mipSize = mip0Size >> 2 * m;
-				offset += mipSize * desc.numMips;
+				unsigned int mipWidth = (desc.width >> m) > 1 ? (desc.width >> m) : 1;
+				unsigned int mipHeight = (desc.height >> m) > 1 ? (desc.height >> m) : 1;
+				unsigned int mipBlocksWide = (mipWidth + desc.blockWidth - 1) / desc.blockWidth;
+				unsigned int mipBlocksTall = (mipHeight + desc.blockHeight - 1) / desc.blockHeight;
+				unsigned long long mipSize = mipBlocksWide * mipBlocksTall * desc.bitsPerPixelOrBlock;
+				offset += mipSize * ((m == mip) ? slice : desc.numMips);
```

Comment:
I made a comment before regarding this but it's fine as it is. If you do the naming changes I'll merge

Reply:
I guess it was in regards to the heavy amount of copy paste? I'm not a big fan either but I didn't feel it was the right call to refactor it too much given the small scope of the library and it would've implicitly added some extra functions that you might not want the class to expose.

I'll do the naming changes.

Reply:
Ah, this line in particular is the 3D Mipmap calculation. It's definitely a bit unusual but the alternative was indeed having to copy a lot of the logic to the final iteration, so this seemed to solve it pretty cleanly. But I agree the logic may not be as obvious to a reader.